core/state/access_events.go

```diff
@@ -0,0 +1,314 @@
+	}
+	if _, hasSelector := ae.chunks[chunkKey]; !hasSelector {
+		chunkRead = true
+		ae.chunks[chunkKey] = AccessWitnessReadFlag
+	}
+
+	// Write access.
+	var branchWrite, chunkWrite, chunkFill bool
+	if isWrite {
+		if (ae.branches[branchKey] & AccessWitnessWriteFlag) == 0 {
+			branchWrite = true
+			ae.branches[branchKey] |= AccessWitnessWriteFlag
+		}
+
+		chunkValue := ae.chunks[chunkKey]
+		if (chunkValue & AccessWitnessWriteFlag) == 0 {
+			chunkWrite = true
+			ae.chunks[chunkKey] |= AccessWitnessWriteFlag
+		}
+		// TODO: charge chunk filling costs if the leaf was previously empty in the state
+	}
+	return branchRead, chunkRead, branchWrite, chunkWrite, chunkFill
+}
+
+type branchAccessKey struct {
+	addr      common.Address
+	treeIndex uint256.Int
+}
+
+func newBranchAccessKey(addr []byte, treeIndex uint256.Int) branchAccessKey {
+	var sk branchAccessKey
+	copy(sk.addr[20-len(addr):], addr)
+	sk.treeIndex = treeIndex
+	return sk
+}
+
+type chunkAccessKey struct {
+	branchAccessKey
+	leafKey byte
+}
+
+func newChunkAccessKey(branchKey branchAccessKey, leafKey byte) chunkAccessKey {
+	var lk chunkAccessKey
+	lk.branchAccessKey = branchKey
+	lk.leafKey = leafKey
+	return lk
+}
+
+// CodeChunksRangeGas is a helper function to touch every chunk in a code range and charge witness gas costs
+func (ae *AccessEvents) CodeChunksRangeGas(contractAddr []byte, startPC, size uint64, codeLen uint64, isWrite bool) uint64 {
+	// note that in the case where the copied code is outside the range of the
+	// contract code but touches the last leaf with contract code in it,
+	// we don't include the last leaf of code in the AccessWitness.  The
+	// reason that we do not need the last leaf is the account's code size
+	// is already in the AccessWitness so a stateless verifier can see that
+	// the code from the last leaf is not needed.
+	if (codeLen == 0 && size == 0) || startPC > codeLen {
+		return 0
+	}
+
+	endPC := startPC + size
+	if endPC > codeLen {
+		endPC = codeLen
+	}
+	if endPC > 0 {
+		endPC -= 1 // endPC is the last bytecode that will be touched.
+	}
+
+	var statelessGasCharged uint64
+	for chunkNumber := startPC / 31; chunkNumber <= endPC/31; chunkNumber++ {
+		treeIndex := *uint256.NewInt((chunkNumber + 128) / 256)
+		subIndex := byte((chunkNumber + 128) % 256)
+		gas := ae.touchAddressAndChargeGas(contractAddr, treeIndex, subIndex, isWrite)
+		var overflow bool
+		statelessGasCharged, overflow = math.SafeAdd(statelessGasCharged, gas)
+		if overflow {
+			panic("overflow when adding gas")
+		}
+	}
+	return statelessGasCharged
+}
+
+// VersionGas returns the amount of gas to be charged if the account's version hasn't been
+// touched in the proper mode. If `isWrite` is `true` then the charged gas is for an access
+// in write mode. If false, the charged gas corresponds to an access in read mode.
+// Note that an access in write mode implies an access in read mode, whereas an access in
+// read mode does not imply an access in write mode.
```

Comment:
From the docs, it sounds like it is a pure function which does a calculation and returns the result. It actually also mutates the underlying access events. 

```suggestion
// VersionGas adds the accounts version to the accessed data, and returns the 
// amount of gas that it costs. 
// Note that an access in write mode implies an access in read mode, whereas an access in
// read mode does not imply an access in write mode.
```
Same goes for all